hir-analysis/src/validation/typecheck.rs

```diff
@@ -0,0 +1,1133 @@
+///
+/// It is separate from the [TypeCheck] rule itself to factor out
+/// all the instruction-related boilerplate.
+struct InstTypeChecker<'a> {
+    diagnostics: &'a DiagnosticsHandler,
+    dfg: &'a DataFlowGraph,
+    span: SourceSpan,
+    opcode: Opcode,
+    pattern: InstPattern,
+}
+impl<'a> InstTypeChecker<'a> {
+    /// Create a new instance of the type checker for the instruction represented by `node`.
+    pub fn new(
+        diagnostics: &'a DiagnosticsHandler,
+        dfg: &'a DataFlowGraph,
+        node: &InstNode,
+    ) -> Result<Self, ValidationError> {
+        let span = node.span();
+        let opcode = node.opcode();
+        let pattern = match opcode {
+            Opcode::Assert | Opcode::Assertz => InstPattern::UnaryNoResult(Type::I1.into()),
+            Opcode::AssertEq => InstPattern::BinaryMatchingNoResult(Type::I1.into()),
+            Opcode::ImmI1 => InstPattern::Unary(Type::I1.into()),
+            Opcode::ImmU8 => InstPattern::Unary(Type::U8.into()),
+            Opcode::ImmI8 => InstPattern::Unary(Type::I8.into()),
+            Opcode::ImmU16 => InstPattern::Unary(Type::U16.into()),
+            Opcode::ImmI16 => InstPattern::Unary(Type::I16.into()),
+            Opcode::ImmU32 => InstPattern::Unary(Type::U32.into()),
+            Opcode::ImmI32 => InstPattern::Unary(Type::I32.into()),
+            Opcode::ImmU64 => InstPattern::Unary(Type::U64.into()),
+            Opcode::ImmI64 => InstPattern::Unary(Type::I64.into()),
+            Opcode::ImmFelt => InstPattern::Unary(Type::Felt.into()),
+            Opcode::ImmF64 => InstPattern::Unary(Type::F64.into()),
+            Opcode::Alloca => InstPattern::Exact(vec![], vec![TypePattern::Pointer]),
+            Opcode::MemGrow => InstPattern::Unary(Type::U32.into()),
+            opcode @ Opcode::GlobalValue => match node.as_ref() {
+                Instruction::GlobalValue(GlobalValueOp { global, .. }) => {
+                    match dfg.global_value(*global) {
+                        GlobalValueData::Symbol { .. } | GlobalValueData::IAddImm { .. } => {
+                            InstPattern::Exact(vec![], vec![TypePattern::Pointer])
+                        }
+                        GlobalValueData::Load { ref ty, .. } => {
+                            InstPattern::Exact(vec![], vec![ty.clone().into()])
+                        }
+                    }
+                }
+                inst => panic!("invalid opcode '{opcode}' for {inst:#?}"),
+            },
+            Opcode::Load => InstPattern::UnaryMap(TypePattern::Pointer, TypePattern::Any),
+            Opcode::Store => {
+                InstPattern::Exact(vec![TypePattern::Pointer, TypePattern::Any], vec![])
+            }
+            Opcode::MemCpy => InstPattern::Exact(
+                vec![TypePattern::Pointer, TypePattern::Pointer, Type::U32.into()],
+                vec![],
+            ),
+            Opcode::PtrToInt => InstPattern::UnaryMap(TypePattern::Pointer, TypePattern::Int),
```

Comment:
We allow casting any integral type to a pointer, so that's why it's `Int` here and not `UInt`, though in general casting signed integers to a pointer is not very common, but there's no reason to disallow it.

Reply:
I should note the reverse is true as well